src/main.rs

```diff
@@ -25,16 +29,29 @@ fn main() -> Result<()> {
         )
         .unwrap();
 
-    let midi_config = config::midi::Midi {
-        port: Some("Launchkey Mini MK3 DAW Port".to_string()),
-    };
-    let midi_messages = match &midi_config.port {
+    let midi_port_names = midi::port_names()?;
+
+    if config.midi.port.is_none() && !midi_port_names.is_empty() {
+        println!(
+            "Config is missing 'midi.port'. Available MIDI ports are:\n{}",
+            midi_port_names.join("\n")
+        );
+    }
+
+    let midi_messages = match &config.midi.port {
         Some(port_name) => Some(midi::listen_for_input(port_name)?),
         None => None,
     };
 
-    let line_launcher: LineLauncher = Progression::parse("C C C C Eb Eb Eb Eb").unwrap().into();
-    line_launcher.listen(beat_message_receiver, conn_out, midi_messages.unwrap());
+    let (progression, duration_ratio_slider) =
+        (config.progression, config.midi.duration_ratio_slider);
```

Comment:
Is this a reasonable way to take ownership of two sub-parts of a struct?

Reply:
I think the more typical way would be to do this:

```rust
let Config { progression, midi: Midi { duration_ration_slider, .. }, .. } = config
```

(I may have the syntax a bit wrong there.)

Reply:
Your way is easier to read, though 🤷🏻

Reply:
Ah ok that makes sense, updated to use destructuring assignment